src/row_codec.rs

```diff
@@ -1,182 +1,230 @@
-                let expire_ts = data.get_i64();
-                meta.expire_ts = if expire_ts == NO_EXPIRE_TS {
-                    None
-                } else {
-                    Some(expire_ts)
-                };
+    pub fn encode(&self, output: &mut Vec<u8>, row: &SstRowEntry) {
+        match &row.key {
+            SstRowKey::SuffixOnly { prefix_len, suffix } => {
+                output.put_u16(*prefix_len as u16);
+                output.put_u16(suffix.len() as u16);
+                output.put(suffix.as_ref());
+            }
+            _ => unreachable!("only suffix only key is supported on encode"),
+        }
+
+        let flags = {
+            let mut flags = RowFlags::empty();
+            if row.expire_ts.is_some() {
+                flags |= RowFlags::HAS_EXPIRE_TS;
+            }
+            if row.create_ts.is_some() {
+                flags |= RowFlags::HAS_CREATE_TS;
+            }
+            if row.value.as_option().is_none() {
+                flags |= RowFlags::Tombstone;
             }
+            flags
+        };
+
+        // encode seq & flags
+        output.put_u64(row.seq);
+        output.put_u8(flags.bits());
+        if flags.contains(RowFlags::Tombstone) {
+            return;
         }
+
+        // encode expire & create ts
+        if flags.contains(RowFlags::HAS_EXPIRE_TS) {
+            output.put_i64(row.expire_ts.unwrap_or(0));
+        }
+        if flags.contains(RowFlags::HAS_CREATE_TS) {
+            output.put_i64(row.create_ts.unwrap_or(0));
+        }
+
+        // encode value
+        let val = row
+            .value
+            .as_option()
+            .expect("value is not set with no tombstone");
+        output.put_u32(val.len() as u32);
+        output.put(val.as_ref());
     }
-    Ok(meta)
-}
 
-fn decode_row_flags(flags: u8) -> Result<RowFlags, SlateDBError> {
-    match RowFlags::from_bits(flags) {
-        None => Err(SlateDBError::InvalidRowFlags),
-        Some(flags) => Ok(flags),
+    pub fn decode(&self, first_key: &Bytes, data: &mut Bytes) -> Result<SstRowEntry, SlateDBError> {
+        let key_prefix_len = data.get_u16() as usize;
+        let key_suffix_len = data.get_u16() as usize;
+        let key_suffix = data.slice(..key_suffix_len);
+        data.advance(key_suffix_len);
+
+        // reconstruct full key
+        let mut full_key = BytesMut::with_capacity(key_prefix_len + key_suffix_len);
+        full_key.extend_from_slice(&first_key[..key_prefix_len]);
+        full_key.extend_from_slice(&key_suffix);
+
+        // decode seq & flags
+        let seq = data.get_u64();
+        let flags = RowFlags::from_bits(data.get_u8()).ok_or(SlateDBError::InvalidRowFlags)?;
+
+        if flags.contains(RowFlags::Tombstone) {
+            return Ok(SstRowEntry {
+                key: SstRowKey::Full(full_key.freeze()),
+                seq,
+                flags,
+                expire_ts: None,
+                create_ts: None,
```

Comment:
Almog had a comment earlier about tombstone having `*_ts` as well.

Reply:
i think `create_ts` is nice to have, but i was afraid of it might less makes sense for tombstone to have an `expire_ts`.. to expire a tombstone, we might have to "recover" the deleted rows with tombstone, it'd be a rare behaviour among the existed kv engines.